selenium_tests/util.py

```diff
@@ -0,0 +1,158 @@
+        ]
+
+    def assert_no_console_errors(self):
+        console_errors = self.get_console_errors()
+        assert len(console_errors) == 0, str(console_errors)
+
+    def num_matching_elements(self, selector):
+        script = "return document.querySelectorAll({selector!r}).length".format(selector=selector)
+        return self.driver.execute_script(script)
+
+    def get(self, relative_url):
+        url = make_absolute_url(relative_url, self.live_server.url)
+        self.driver.get(url)
+        self.wait_until_loaded(By.TAG_NAME, 'body')
+        self.assert_no_console_errors()
+
+    def set_dimension(self, width=None, height=None):
+        """Helper function to set the width or height, or both at once"""
+        current_dimensions = self.driver.get_window_size()
+        if width is None:
+            width = current_dimensions['width']
+        if height is None:
+            height = self.driver.execute_script("return document.body.scrollHeight")
+            # height = current_dimensions['height']
+        self.driver.set_window_size(width, height)
+
+    def take_screenshot(self, name=None, output_base64=False):
+        self.set_dimension()
+        if name is None:
+            name = now_in_utc().strftime('%Y_%m_%d_%H_%M_%S_%f')
+        repo_root = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+        filename = os.path.join(repo_root, "{}.png".format(name))
+        self.driver.save_screenshot(filename)
+        print("PNG screenshot for {name} output to {filename}".format(
+            name=name,
+            filename=filename,
+        ))
+        if output_base64:
+            # Can be useful for travis where we don't have access to build artifacts
+            with open(filename, 'rb') as f:
+                print("Screenshot as base64: {}".format(b64encode(f.read())))
+
+
+def make_absolute_url(relative_url, absolute_base):
+    """
+    Create an absolute URL for selenium testing given a relative URL. This will also replace the host of absolute_base
+    with the host IP of this instance.
+
+    Args:
+        relative_url (str): A relative URL
+        absolute_base (str): An absolute URL which contains the http port and scheme we need
+
+    Returns:
+        str: An absolute URL pointing to the live server instance
+    """
+    # Swap out the hostname, which was set to 0.0.0.0 to allow external connections
+    # Change it to use ip of this container instead
+    absolute_pieces = urlparse(absolute_base)
+    relative_pieces = urlparse(relative_url)
+    host = socket.gethostbyname(socket.gethostname())
+    absolute_url = ParseResult(
+        absolute_pieces.scheme,
+        "{host}:{port}".format(host=host, port=absolute_pieces.port),
+        relative_pieces.path,
+        relative_pieces.params,
+        relative_pieces.query,
+        relative_pieces.fragment,
+    ).geturl()
+    if '?q' in relative_url:
```

Comment:
i'll remove this at some point. for some reason, loading the url with the querystring was being immediately cut down to just the url, but _only_ when i ran the entire suite. when i run just the single test case, it works fine. still need to figure this out